### PR TITLE
feature/string-attributes: fixed issue#12

### PR DESCRIPTION
**Pull request by kcning**
Hi, I finished the missing part! #12

### CODE REVIEW

File: lib/weka/core/attribute.rb [hunk 1]
```diff
@@ -70,6 +76,17 @@ def internal_value_of(value)
...
+        if string?
+          # string has unlimited range of possible values. We check if
+          # it has been added. If yes then we simply return the index
+          if (idx = index_of_value(value.to_s)) != -1
+            idx
+          else
+            # else we add the value to the range and return its new index
+            add_string_value(value.to_s)
+          end
+        end
```
Comment (paulgoetze) [inline lib/weka/core/attribute.rb]: I think the string handling could be moved into a separate private method, in order to keep `#internal_value_of` as concise as possible. Maybe sth. like `index_of_string(value)`?

With this, we could have a one-liner as for the other types (I'd use the `return` here as well, even if it is not necessary):

```ruby
return index_of_string(value) if string?
```

I would also avoid assignments in if clauses and would rather do it before with a fully readable name, like:

```ruby
index = index_of_value(value.to_s)

if index == -1
  # ...
else
 # ...
end
```

The comments could then be the method’s comment and move out of the body (Should work, right? The new method should be short enough then, to get what it refers to)

Could you make these changes? Then we are good to go!

**Comment by kcning** [comment]
Oke now we have the spec (thanks to it a bug is found). Can you please check?

File: lib/weka/core/instances.rb [hunk 2]
```diff
@@ -226,6 +226,16 @@ def instance_from(instance_or_values, weight:)
...
+          # string attribute has unlimited range of possible values.
+          # Check the return index, if it is -1 then add the value to
+          # the attribute before creating the instance
+          data.map!.with_index do |value, index|
+            if value == -1 and attribute(index).string?
+              attribute(index).add_string_value(instance_or_values[index].to_s)
+            else
+              value
+            end
+          end
```
Comment (paulgoetze) [inline lib/weka/core/instances.rb]: You are right–adding the string value should not be done in the Attribute’s `#internal_value_for` but already during the creation of the instance. Nice catch!

**Comment by paulgoetze** [comment]
Tests are still missing for the implementation of adding string values–all tests pass if you comment out the added code (https://github.com/paulgoetze/weka-jruby/pull/13/files#diff-bb64f986c10a4d064a282d726d7b6c77R232). 
Anyway, I also found some other missing things to fully support string attributes.
 
So, I will just go ahead and merge this and do the rest of the changes in #12 (I think the easiest is to add a string attribute to the test dataset files to cover all string cases in the tests as well).

Thanks again for your contributions!

**Comment by kcning** [comment]
Thanks! I'm using the gem for some string data mining so I can add the use cases to the spec. If I come across the missing methods for string attributes I can also add it. (What are the missing things, actually?)

**Comment by paulgoetze** [comment]
Nice, then go ahead, I really appreciate some help here!

One missing thing I came across is `DenseInstance#value_from` needs string handling, too:
https://github.com/paulgoetze/weka-jruby/blob/b44ee9b786b85a9b70212e6fa489a4fc1a87d6a8/lib/weka/core/dense_instance.rb#L54

I will have a look and let you know, when I spot other stuff.

One other thing: could you make sure your code is compliant with default [Rubocop](https://github.com/bbatsov/rubocop) recommendations? I used it for this project, yet need to add a hint to the Readme.

Further discussion can happen either in #12 or a new PR. Thanks!

**Pull request merged**